### PR TITLE
Improve and tighten pure functions specification

### PR DESCRIPTION
**Pull request by andralex**
This is in keeping with the current implementation and tightens the spec to ensure no future shenanigans.

### CODE REVIEW

File: spec/function.dd [hunk 1]
```diff
@@ -224,6 +222,26 @@ pure int foo(int i,
...
+    $(P Any `pure` function that is not strongly pure cannot be memoized. The
+    compiler is required to honor all calls to the function, even if it appears
+    to do nothing. (Example: a pure function taking no arguments and returning
+    `int*` cannot be memoized. It also should not because it may be a typical
+    factory function returning a fresh pointer with each call.))
+
+    $(P `pure` functions returning `void` will be always called even if it is
+    strongly `pure`. The implementation must assume the function does something
+    outside the confines of the type system and is therefore not allowed to
+    elide the call, even if it appears to have no possible effect.)
```
Comment (ZombineDev) [inline spec/function.dd]: Wouldn't it be simpler to just say that `pure` functions returning `void` are not strongly `pure`?

Comment (ZombineDev) [inline spec/function.dd]: That's sort of implied by "... (b) returns a result without mutable indirections", if we agree that `void` functions don't return any result.

**Comment by WalterBright** [comment]
Should cover what happens when an exception is thrown by the pure function. I suggest regarding the exception thrown as its 'return value' as far as uniqueness and memoization rules go.

**Comment by andralex** [comment]
@WalterBright updated with no special requirements for exceptions so as to leave maximum freedom

**Comment by andralex** [comment]
@WalterBright ping - guess we're now good

File: spec/function.dd [hunk 2]
```diff
@@ -224,6 +228,51 @@ pure int foo(int i,
...
+
+    $(P Any `pure` function that is not strongly pure cannot be assumed to be
+    memoizable and enjoys no special treatment compared to a regular non-pure
+    function. (Example: a pure function taking no arguments and returning `int*`
+    should not be memoized. It also should not because it may be a typical
+    factory function returning a fresh pointer with each call.) Function calls
+    may still be elided, or results be memoized, by means of traditional
+    inlining and optimization techniques available for all functions.)
+
+    $(P A function that accepts only parameters without mutable indirections and
```
Comment (JohanEngelen) [inline spec/function.dd]: whoops, should be: "A _`pure`_ function" ?

**Comment by andralex** [comment]
@JohanEngelen thx fixed

File: spec/function.dd [hunk 3]
```diff
@@ -271,6 +271,9 @@ pure List* make(int a, int b)
     assumptions related to memoization and references do not carry to the thrown
     exception.)
 
+    $(P Destructors will always be called even if they appear to be strongly
```
Comment (JohanEngelen) [inline spec/function.dd]: "Destructors will always be called" is not want you want to say I think; instead, it should only say something about pure destructors.
Perhaps: "When a destructor is pure, it does not imply that it need not be called."


Maybe I am approaching this from too low-level viewpoint. When you write "destructors will always be called", do you mean it in the sense of debug-call-tracing-low-levelness, or in the high level sense of "the program will be executed as if destructors are always called" (thus allowing empty destructor calls to be eliminated)?

**Comment by nordlow** [comment]
Can we go further with make free and GC add/remove Range pure now?

**Pull request merged**